src/Mod/Part/Gui/ViewProvider2DObject.cpp

```diff
@@ -91,11 +92,28 @@ ViewProvider2DObjectGrid::~ViewProvider2DObjectGrid()
 // **********************************************************************************
 
 SoSeparator* ViewProvider2DObjectGrid::createGrid()
+{
+    Gui::coinRemoveAllChildren(GridRoot);
+    int numberOfLineIfGridAuto = 200; //fills the view in most cases.
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Part");
+    int numberSubdivision = hGrp->GetInt("GridNumberSubdivision", 10);
+
+    //First we create the subdivision lines
+    createGridPart(GridSize.getValue(),
+        hGrp->GetInt("GridLinePattern", 0x0f0f), numberOfLineIfGridAuto, hGrp->GetInt("GridSmallLineWidth", 1));
+
+    //Second we create the wider lines marking every nth lines
+    createGridPart(GridSize.getValue() * numberSubdivision,
+        hGrp->GetInt("GridTenthLinePattern", 0xffff), numberOfLineIfGridAuto / numberSubdivision, hGrp->GetInt("GridBigLineWidth", 2));

```

Comment:
I did this like this initially because it was easier to test. But I think this createGrid function needs to be rewriten completely.

The reason is that currently we are creating a 2D object around the origin while all we really need is a grid on the current 3dView. Because the grid is centered on the origin, we end up with suboptimal solutions:
- Limited area of the grid. 
- Grid requiring resizing when the user draw something out of current grid area.
- 10 000 grid line limit. There should be no need to draw 10k lines ever in the first place. Even on a 4K display, 10k lines should cover the whole display. I recall a forum post recently complaining that the 10k grid line limit was not working anymore or something. From which I understand that some people are actually using enormous amount of lines.

However before rewriting it, I wanted to understand if the grid is actually used by something else than sketcher.

Reply:
the huge 10k limit is probably there precisely because the grid is centered around the origin, this allows to have objects at a certain distance from origin but still have a grid, I agree ideally the grid should only be drawn on the visible 3d view and that 10k is absurd for that, however having the grid only on the visible area would mean that it would have to be updated all the time on zoom or pan, which might not be ideal, perhaps it would be a better idea to have the grid on something like 10 times the visible size or something.

Reply:
Yes I was thinking about it.
The best solution I found to not recreating the grid a billion times, was that the grid should be 3 times the size of the viewer. And it should redraw only when user stop paning. ie when he release the click and the view settles (and on zoom).
Because user can't pan more than the size of the viewer. So if grid is 3 times the viewer size, then he can't pan out of the grid.